### PR TITLE
Convert OFT to LoRA

### PR DESCRIPTION
**Pull request by dxqb**
This is prototype code to convert any OneTrainer-compatible Peft to LoRA
It doesn't work for OFT though - please see comments

### CODE REVIEW

File: modules/module/LoRAModule.py [hunk 1]
```diff
@@ -333,6 +350,26 @@ def extract_from_module(self, base_module: nn.Module):
         # TODO
         pass
 
+    def get_applied_weights(self, device: torch.device) -> Tensor:
```
Comment (dxqb) [inline modules/module/LoRAModule.py:353]: this function is used when the source is a LoRA (LoRA-to-LoRA conversion). this merges the original module weights with the LoRA weights

all the following code works then, and sampling creates an almost identical image even though the LoRA weights have gone through SVD

File: modules/module/LoRAModule.py [hunk 2]
```diff
@@ -430,6 +467,24 @@ def initialize_weights(self):
 
         nn.init.zeros_(self.oft_R.weight)
 
+    def get_applied_weights(self, device: torch.device) -> Tensor:
+        #taken from Conv2d code, assuming it works the same for linear:
```
Comment (dxqb) [inline modules/module/LoRAModule.py:505]: if the source is an OFT though, the resulting sample is almost the same sample as generated from the untrained model. The LoRA converted from OFT doesn't have an effect.

does this code also apply to Linear layers, or only to Conv2d? Could you add code that applies to Linear layers?

File: modules/module/LoRAModule.py [hunk 3]
```diff
@@ -430,6 +467,24 @@ def initialize_weights(self):
...
+        orth_rotate = self.oft_R._cayley_batch(
```
Comment (dxqb) [inline modules/module/LoRAModule.py:506]: the result from this seems to be an identity matrix

File: modules/module/LoRAModule.py (hunk 2 above)
Comment (Koratahiu) [reply modules/module/LoRAModule.py:505]: Yeah, this only applies for Conv2d, since Transformer models have only linear layers it's expected to not work.

File: modules/module/LoRAModule.py (hunk 3 above)
Comment (Koratahiu) [reply modules/module/LoRAModule.py:506]: I think the check for linear should be here, something like what I did in my comfy implementation:
```
             if is_conv2d:
                # Use R for Conv2d layers
                rotation_matrix_for_weight = orth_rotate
            else:
                # Use R.T for Linear layers
                rotation_matrix_for_weight = orth_rotate.transpose(-1, -2)
```

File: modules/module/LoRAModule.py [hunk 4]
```diff
@@ -430,6 +467,24 @@ def initialize_weights(self):
...
+            self.oft_R.weight, self.oft_R.block_size, self.oft_R.use_cayley_neumann, self.oft_R.num_cayley_neumann_terms
+        )
+        #TODO probably shouldn't use dropout for this, only for training?
+        orth_rotate = self.oft_R.dropout(orth_rotate)
```
Comment (Koratahiu) [inline modules/module/LoRAModule.py:510]: OFT dropout is different. It forces random nodes to revert to the original model starting point (I didn't test it tbh), so it's not a good idea to add it here?

File: modules/module/LoRAModule.py (hunk 2 above)
Comment (dxqb) [reply modules/module/LoRAModule.py:505]: I see. Could you implemented `get_applied_weights` in a way that works both for Linear and Conv2d? I cannot.

The way Linears are handled in forward() is good if you have activations, but for conversion we need the full matrix.

Comment (Koratahiu) [reply modules/module/LoRAModule.py:505]: I'll look into it
